posts/2019-01-14-easy-peasy-nix-versions.md

```diff
@@ -0,0 +1,175 @@
+title: Easy Peasy Nix Versions
+---
+
+<style>
+.story {
+  background-color: lightblue;
+  padding: 20px;
+  border-radius: 10px;
+}
+
+</style>
+
+# Easy Peasy Nix Versions
+
+This is a convention for using third-party packages in Nix. It has a simple
+directory structure, makes using the packages straightforward and automatizes
+updates.
+
+---
+
+Where should you declare the third-party packages you use in your Nix project?
+Should all specs be declared in the `default.nix`? Should each third-party
+package get its own directory with a `default.nix` and a `spec.src.json`? Or
+maybe the specs should only exist at call site, potentially duplicated if the
+package is used in different places? I settled on a simple convention which
+I've been using for a year across all my personal and work Nix projects.
+
+The idea is to use a single JSON file to store all versions rather than having
+each package in a directory defining its URL and version separately (I tend to
+get lost with the latter). Two more files are involved: one that acts as glue
+between JSON and Nix, and one that automatizes the package updates. It's not a
+full-blown solution like [require.nix] but it does the job very well for small-
+to medium-sized project.
+
+## Specifying and fetching third-party packages
+
+We'll take the [homies] repository as an example.
+
+<div class="story">
+[homies] is a Nix-based reproducible environment, you can read about it in the
+[dedicated article][homies-article].
+</div>
+
+If you checkout the repository you'll see a `nix/` directory with two files:
+`versions.json` and `fetch.nix`.  The [`nix/versions.json`] file contains
+information about _where_ to find each of the third-party packages:
+
+``` json
+{
+  "nixpkgs": {
+    "owner": "NixOS",
+    "repo": "nixpkgs-channels",
+    "branch": "nixos-18.09",
+    "rev": "9d608a6f592144b5ec0b486c90abb135a4b265eb",
+    "sha256": "03brvnpqxiihif73agsjlwvy5dq0jkfi2jh4grp4rv5cdkal449k"
+  },
+  "snack": {
+    "owner": "nmattia",
+    "repo": "snack",
+    "branch": "master",
+    "rev": "9754f4120d28ce25888a608606deddef9f490654",
+    "sha256": "1a2gcap2z41vk3d7cqydj880lwcpxljd49w6srb3gplciipha6zv"
+  }
+}
+```
+
+Each third-party package is given a name, like `nixpkgs` or `snack` in the
+above. That name is used as a key in a JSON dictionary. For each package the
+corresponding value is the GitHub repository information as well as the
+`sha256` for `fetchurl` (this is tailored for GitHub, but make sure to read the
+[closing thoughts](#closing-thoughts) for ideas). The `branch` attribute is not
+used by Nix but comes in handy for [updating the
+packages](#updating-third-party-packages).
+
+The [`nix/fetch.nix`] file is a wrapper for using the third-party packages in
+your Nix code. The third-party packages are retrieved using the name you gave
+them in the JSON file; see the [`default.nix`] in [homies]:
+
+``` nix
+with { fetch = import ./nix/fetch.nix; };
+let
+  pkgs = import (fetch "nixpkgs") {};
```

Comment:
`fetch.nixpkgs` might be best if the reference is used multiple times. That way the thunk is only resolved once. Since nix is lazy it shouldn't make much difference.

Reply:
oh nice :ok_hand: